pkg/server/service.go

```diff
@@ -262,8 +262,21 @@ func (c *criContainerdService) Run(startGRPC bool) error {
 
 	<-eventMonitorCloseCh
 	logrus.Info("Event monitor stopped")
-	<-streamServerCloseCh
-	logrus.Info("Stream server stopped")
+	// There is a race condition with http.Server.Serve.
+	// When `Close` is called at the same time with `Serve`, `Close`
+	// may finish first, and `Serve` may still block.
+	// See https://github.com/golang/go/issues/20239.
+	// Here we set a 2 second timeout for the stream server wait,
+	// if it timeout, an error log is generated.
+	// TODO(random-liu): Get rid of this after https://github.com/golang/go/issues/20239
+	// is fixed.
+	const streamServerStopTimeout = 2 * time.Second
```

Comment:
if we called serve after close, they are deadlocked and the amount of time we wait for close is moot...

It seems from reading the golang issue that that we should not be calling serve after close and the golang mitigation will be to give us an error if we do. Maybe wrap it and stop ourselves from serve after close?

Reply:
> if we called serve after close, they are deadlocked and the amount of time we wait for close is moot...

1) In most cases, `Close` is called after `Serve`, in that case we do want to wait for stream server to stop here.
2) If we happen to hit the deadlock, there is no way to figure it out. The only solution is to add a deadlock here.

> It seems from reading the golang issue that we should not be calling serve after close and the golang mitigation in 1.11 will be to give us an error if we do.

There are 2 problems:
1) Even we add check before `ListenAndServe`, it's still possible that `Close` is called after `ListAndServe` is called but before the internal `Serve`. There is still race.
2) Even with the proposed golang fix, I still don't think it completely migrate the race condition. `Close` can still be called after `Serve` is called, but before `trackListener`.

I'll also comment there.

Reply:
Someone does mentioned what I said:
```
I don't think this fixes the original bug. What happens if the Shutdown call happens between line 2726 (Serve calls shuttingDown) and line 2739 (Serve calls trackListener)? We should add the original example from #20239 as a test: https://play.golang.org/p/ESmgJ6O_WU
```

Reply:
> If we happen to hit the deadlock, there is no way to figure it out. 

Can we use a mutex to manage the call sequence via state progression/validation and to remove the race condition?

Reply:
No. Unless we can add mutex inside golang function.